File: 110. Balanced Binary Tree-1.cpp
class Solution {
public:
  bool res = true;
  int isBalanced(TreeNode *root) {
    helper(root);

    return res;
  }

  int helper(TreeNode *root) {
    if (!root)
      return 0;

    int l = helper(root->left);
    int r = helper(root->right);

    if (abs(l - r) > 1)
      res = false;

    return max(l, r) + 1;
  }
};

class Solution {
public:
  bool isBalanced(TreeNode *root) { return dfs(root) < 0 ? false : true; }

  int dfs(TreeNode *root) {
    if (!root) {
      return 0;
    }
    int l = dfs(root->left);
    int r = dfs(root->right);
    if (l == -1 || r == -1)
      return -1;
    if (abs(l - r) > 1) {
      return -1;
    }

    return max(l, r) + 1;
  }
};

File: 110. Balanced Binary Tree-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def f(root):
            if root is None:
                return 0
            left = f(root.left)
            if left < 0:
                return -1
            right = f(root.right)
            if right < 0 or abs(left - right) > 1:
                return -1
            return max(left, right) + 1

        return f(root) != -1
